src/api.proposed.x.d.ts

```diff
@@ -0,0 +1,171 @@
+         * Returns the sub-protocols to be used. See details of `protocols` here https://developer.mozilla.org/en-US/docs/Web/API/WebSocket/WebSocket
+         * Useful if there is a custom authentication scheme that needs to be used for WebSocket connections.
+         * Note: The client side npm package @jupyterlab/services uses WebSockets to connect to remote Kernels.
+         */
+        readonly webSocketProtocols?: string[];
+    }
+
+    /**
+     * Represents a Jupyter Server displayed in the list of Servers.
+     */
+    export interface JupyterServer {
+        /**
+         * Unique identifier for this server.
+         */
+        readonly id: string;
+        /**
+         * A human-readable string representing the name of the Server. This can be read and updated by the extension.
+         */
+        label: string;
+    }
+
+    /**
+     * Provider of Jupyter Servers.
+     */
+    export interface JupyterServerProvider {
+        /**
+         * Event fired when the list of servers changes.
+         * Note: the getJupyterServers method will not be called unless changes are detected.
+         */
+        onDidChangeServers?: Event<void>;
+        /**
+         * Returns the list of servers.
+         */
+        getJupyterServers(token: CancellationToken): Promise<JupyterServer[]>;
+        /**
+         * Returns the connection information for the Jupyter server.
+         */
+        resolveConnectionInformation(
+            server: JupyterServer,
+            token: CancellationToken
+        ): Promise<JupyterServerConnectionInformation>;
+    }
+    /**
+     * Represents a reference to a Jupyter Server command.
+     */
+    export interface JupyterServerCommand {
+        /**
+         * Title of the command, like `save`.
+         */
+        title: string;
+        /**
+         * A human-readable string which is rendered less prominent in a separate line.
+         */
+        detail?: string;
+        /**
+         * A tooltip for the command, when represented in the UI.
+         */
+        tooltip?: string;
+        /**
+         * Default command to be used when there are no servers.
+         * If not set, and there are not servers, then the user will be prompted to select a command from a list of commands returned by `getCommands`.
+         */
+        picked?: boolean;
+    }
+    /**
+     * Provider of Jupyter Server Commands.
+     * Each command allows the user to perform an action.
+     */
+    export interface JupyterServerCommandProvider {
+        /**
+         * Returns a list of commands to be displayed to the user.
+         * If there are no JupyterServers and one of the returned commands has `picked = true`,
+         * then the `handleCommand` method will be invoked with that command.
+         * @param options Reserved for future use. Extensions can ignore this argument for now.
+         */
+        getCommands(options: unknown, token: CancellationToken): Promise<JupyterServerCommand[]>;
```

Comment:
@jrieken @mjbvz 
We have a proposed API on top of this.
the Proposed API is only for use by the Jupyter extnsion and another extension that will be similar to Jupyter extension.

The first argument will be `value:string`, which represents the value typed by the user in the quick pick, that
allows the 3rd party extension to change the contributed commands based on the values entered in the quik pick

Basically we're after the ability to dynamically populate quick pick items, just as we do with `Open Remote Repository` quick pick, see screenshots below.

Once again, the proposed API is only for internal use, right now we do not plan on exposing this for others.
However if we did, want to (@mjbvz suggested that it might be useful for other extension), then we have the ability to do so and have a clean api. 
All this to ensure the `token` is the last argument

<img width="622" alt="Screenshot 2023-08-15 at 20 56 22" src="https://github.com/microsoft/vscode-jupyter/assets/1948812/9fff82e4-e808-450d-a995-26a638783080">
<img width="623" alt="Screenshot 2023-08-15 at 20 56 28" src="https://github.com/microsoft/vscode-jupyter/assets/1948812/5f774976-60f9-4252-a561-9e187fd57c2a">

Reply:
> Basically we're after the ability to dynamically populate quick pick items, just as we do with Open Remote Repository quick pick, see screenshots below.

I believe a difference is that the remote repo picker can not list all remotes, that's why it reacts to what's being typed (so that it can query GH as you type). Do you envision a similar use-case and is that generic to all commands or just one?